Fix useMultiline store race condition

finkef opened the pull request:
Alternative implementation of #2077.

Fixes #2041, #2049.

This PR fixes multiline calculation by making sure useLayoutEffect is not executed out-of-sync with React's scheduler and instead only triggered by hook dependencies. The performance seems to not be effected by this change.

raineorshine commented on src/components/Editable/useMultiline.ts:
```diff
@@ -1,17 +1,24 @@
   const showSplitView = useSelector(state => state.showSplitView)
   const splitPosition = useSelector(state => state.splitPosition)
 
+  // Check if the cursor is active.
+  // NOTE: We don't want to select the entire cursor path to skip invoking the layout effect too often.
+  const cursorActive = useSelector(state => equalPath(state.cursor, simplePath))
```
`cursorActive` is not quite adequate to cover the behavior of the `useSelectorEffect`. We need to remeasure whenever the cursor changes, even if the cursor is not active on this thought. In the following example, if the cursor is moved from `aaaaaaaaaa` to `bbbbbbbbbb', then `cccccccccccc` may wrap and become multiline.

```
- aaaaaaaaaa
  - bbbbbbbbbb
    - cccccccccccc
```

finkef commented:
Moved all layout calculations to `useLayoutEffect` and rewrote react ministore's `useSelector` to use `useSyncExternalStore`. After many times going back and forth of trying different store subscription techniques for trying to reduce subscription invocations, it turns out that React's internal comparisons inside `useSyncExternalStore` easily perform best while keeping the code at a minimum.

raineorshine reviewed (changes requested):
This looks great, and is definitely the right integration for the editingValue.

See comments regarding cursor change. Thanks!

raineorshine commented:
There are two very closely related test cases that are slightly different than #2041 and #2049. I created separate issues for them for clarity, but wanted to bring them to your attention now since they involve the same height/line-height rendering calculations.

- #2084 
- #2085

If they are not automatically fixed with one of your PR's for our existing issues, then we can add them to the budget.

finkef commented:
> There are two very closely related test cases that are slightly different than #2041 and #2049. I created separate issues for them for clarity, but wanted to bring them to your attention now since they involve the same height/line-height rendering calculations.
> 
> * [Multiline URL height is incorrect when cursor is set to null #2084](https://github.com/cybersemics/em/issues/2084)
> * [Multiline URL height is incorrect after refresh and set cursor #2085](https://github.com/cybersemics/em/issues/2085)
> 
> If they are not automatically fixed with one of your PR's for our existing issues, then we can add them to the budget.

Thanks, will check them out. On first glance, I'm guessing they should be fixed with the PRs already.

finkef commented on src/components/Editable/useMultiline.ts:
```diff
@@ -11,6 +9,10 @@ const useMultiline = (contentRef: React.RefObject<HTMLElement>, simplePath: Simp
   const fontSize = useSelector(state => state.fontSize)
   const showSplitView = useSelector(state => state.showSplitView)
   const splitPosition = useSelector(state => state.splitPosition)
+  const cursor = useSelector(state => state.cursor)
```
I was afraid that this would cause some out-of-sync issues or race conditions again, but after excessive testing, couldn't find anything. Moved it back to `useSelectorEffect`.

finkef commented on src/components/VirtualThought.tsx:
```diff
@@ -160,13 +157,9 @@ const VirtualThought = ({
     })
   }, [crossContextualKey, onResize, path, thought.id])
 
-  // Read the element's height from the DOM on cursor change and re-render with new height
-  // shimHiddenThought will re-render as needed.
-  useSelectorEffect(updateSize, selectCursor, shallowEqual)
-
```
We're running multiline calculations whenever the cursor changes (which should be the only height-affecting effect here?) and trigger `updateSize` in the resulting `useLayoutEffect`, so I think adding the calculation here too adds redundancy and potential race conditions when conflicting with `useLayoutEffect`.

raineorshine replied:
`updateMultiline` is triggered when the cursor changes, but how does that lead to `updateSize` being called?

finkef replied:
It results in `updateSize` being called here by a change in multiline: https://github.com/cybersemics/em/blob/d188a1309f8f630587c772e60b94a032492a0180/src/components/StaticThought.tsx#L112-L114

But you are right, this wouldn't catch all the cases, e.g. a thought wrapping across 3 lines instead of 2 after a cursor change. Adding it back. 👍

raineorshine commented on src/components/Editable/useMultiline.ts:
```diff
@@ -26,21 +32,17 @@ const useMultiline = (contentRef: React.RefObject<HTMLElement>, simplePath: Simp
 
+  // Recalculate multiline on mount, when the font size changes, edit, split view resize, value changes, and when the
+  // cursor changes to or from the element.
+  useLayoutEffect(() => {
+    updateMultiline()
```
Prefer point-free style here:

```
useLayoutEffect(updateMultiline, ...)
```

finkef replied:
Great call, fixed! 🙌

finkef commented:
@raineorshine Resolved all comments. 👍

finkef commented:
> There are two very closely related test cases that are slightly different than #2041 and #2049. I created separate issues for them for clarity, but wanted to bring them to your attention now since they involve the same height/line-height rendering calculations.
> 
> * [Multiline URL height is incorrect when cursor is set to null #2084](https://github.com/cybersemics/em/issues/2084)
> * [Multiline URL height is incorrect after refresh and set cursor #2085](https://github.com/cybersemics/em/issues/2085)
> 
> If they are not automatically fixed with one of your PR's for our existing issues, then we can add them to the budget.

@raineorshine Can confirm that both of these are fixed by #2079 that is dependent on this PR.

raineorshine reviewed (approved):
Looks great! Thank you.
